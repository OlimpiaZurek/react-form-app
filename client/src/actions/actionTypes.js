export const GET_FORM_DATA = 'GET_FORM_DATA';
export const SEND_FORM_DATA = 'SEND_FORM_DATA';

export const SEND_DATA_SUCCESS = 'SEND_DATA_SUCCESS';
export const SEND_DATA_FAILURE = 'SEND_DATA_ERROR';
