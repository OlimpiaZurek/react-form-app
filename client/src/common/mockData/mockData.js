export const data = {
  firstname: 'test',
  lastname: 'test',
  email: 'test@test.com',
  date: '02.11.2019',
};