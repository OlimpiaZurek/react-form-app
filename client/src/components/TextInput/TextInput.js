import React  from 'react';
import PropTypes from 'prop-types';
import styles from './TextInput.module.scss';

const TextInput = ({
  onClick,
  handleChange, 
  placeholder, 
  value,
  type,
  name,
  className,
  error,
}) => (
  <div onClick={onClick} className={styles.Wrapper}>
    <input 
      type={type}
      value={value} 
      onChange={handleChange} 
      placeholder={placeholder}
      name={name}
      className={styles.Input}
      required
    />
    <div className={styles.Error}>{error}</div>
  </div>
);

TextInput.propTypes = {
  onClick: PropTypes.func,
  handleChange: PropTypes.func,
  placeholder: PropTypes.string,
  value: PropTypes.string,
  type: PropTypes.string,
  name: PropTypes.string,
  className: PropTypes.string,
  error: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.bool
  ]),
};

export default TextInput;
