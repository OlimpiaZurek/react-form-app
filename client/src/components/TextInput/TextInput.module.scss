@import '../../styles/global.scss';

.Input {
    margin: $metrics-xs 0;
    padding: $metrics-s;
    width: 100%;
}

.Error {
    color: $danger-color;
}