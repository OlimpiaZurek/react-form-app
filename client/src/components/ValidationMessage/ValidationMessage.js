import React from 'react';
import PropTypes from 'prop-types';
import styles from './ValidationMessage.module.scss';

const ValidationMessage = ({ status, text }) => (
  <div className={`${styles.Message} 
    ${status ? styles.Success : styles.Error}`}
  >
    {text}
  </div>
);

ValidationMessage.propTypes = {
  status: PropTypes.string,
  text: PropTypes.string,
};

export default ValidationMessage;

