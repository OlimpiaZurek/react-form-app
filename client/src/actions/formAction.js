import * as actionTypes from './actionTypes';
import { sendFormData } from '../services/formService';

export const sendFormDataAction = data => ({
  type: actionTypes.SEND_FORM_DATA,
  data
});

export const sendFormSuccess = message => ({
  type: actionTypes.SEND_DATA_SUCCESS,
  message
});

export const sendFormFailure = message => ({
  type: actionTypes.SEND_DATA_FAILURE,
  message
})

export const submitFormData = data => dispatch =>
  sendFormData(data)
  .then(({ data } )=> {
    dispatch(sendFormDataAction(data))
    if (data.success) {
      dispatch(sendFormSuccess(data.message));
    } else {
      dispatch(sendFormFailure(data.message));
    }
  })
  .catch((err) => {
    dispatch(sendFormFailure(err));
});
