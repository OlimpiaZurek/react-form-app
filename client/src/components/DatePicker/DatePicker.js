import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { debounce } from 'lodash';
import TextInput from '../TextInput/TextInput';
import Calendar from '../Calendar/Calendar';
import { 
  DATE_PICKER_PLACEHOLDER
} from '../../common/consts/calendar';

class DatePicker extends Component {

  constructor(props) {
    super(props);
      this.state = {
        isCalendarOpened: false,
    };
    this.delayedFunc = debounce(this.toggleCalendar, 300)
  }
 
  toggleCalendar = () => {
    this.setState({
      isCalendarOpened: !this.state.isCalendarOpened
    });
  }
  
  handleChange = (e) => {
    e.preventDefault();
  }

  onClick = (day) => {
    this.delayedFunc()
    this.props.onDateClick(day)
  }

  render() {
    const { isCalendarOpened } = this.state;
    const { value } = this.props;
    return (
      <Fragment>
        <TextInput
          type="text"
          value={value}
          placeholder={DATE_PICKER_PLACEHOLDER}
          handleChange={this.handleChange}
          onClick={this.toggleCalendar}
          {...this.props}
        />
        {isCalendarOpened && (
          <Calendar 
            {...this.props}
            onDateClick={this.onClick}
          />
        )}
      </Fragment>
     
    );
  }
}

DatePicker.propTypes = {
 value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.instanceOf(Date)
 ]),
 onDateClick: PropTypes.func,
};

export default DatePicker;
