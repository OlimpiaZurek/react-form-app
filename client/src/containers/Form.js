import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { format } from 'date-fns';
import TextInput from '../components/TextInput/TextInput';
import DatePicker from '../components/DatePicker/DatePicker' ;
import ValidationMessage from '../components/ValidationMessage/ValidationMessage';
import Button from '../components/Button/Button';
import STATE, { 
  reduceFormValues, 
  checkAllFieldsValid 
} from '../common/utils/validation';
import { 
  INPUT_DATE_FORMAT
} from '../common/consts/calendar';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as actions from '../actions/formAction';

const txtFieldState = {
  value: '',
  valid: true,
  typeMismatch: false, 
};

export class EventForm extends Component {

  state = {
    currentDate: new Date(),
    ...STATE(txtFieldState)
  }

  handleChange = (e) => {
    const { name, value } = e.target;
    let newState = Object.assign({}, this.state);
    newState[name].value = value;
    this.setState(newState);
  }

  handleSubmit = (e) => {
    e.preventDefault();
    const form = e.target;
    const formValues = reduceFormValues(form.elements, this.state);
    const allFieldsValid = checkAllFieldsValid(formValues);

    this.setState({ 
      ...formValues,
      allFieldsValid 
    });
    const { 
      firstname, 
      lastname, 
      email, 
      datefield,
    } = formValues;
  
    const preparedData  = {
      firstName: firstname.value,
      lastName: lastname.value,
      email: email.value,
      date: datefield.value
    };


    if (allFieldsValid) {
      this.props.actions.submitFormData(preparedData);
    }
  }

  onDateClick = day => {
    this.setState(prevState => ({
      datefield: {
        ...prevState.datefield,
        value: day
      },
      currentDate: day
    }))
  };
  
  render() {
    const { 
      email,
      firstname,
      lastname,
      datefield, 
      allFieldsValid 
    } = this.state;

    const isMatch = email.typeMismatch && email.formatErrorTxt;
    const isRequired = !email.valid && email.requiredTxt;

    if (allFieldsValid) {
      return (
        <ValidationMessage 
          text={this.props.message}
          status={this.props.status}
        />
      );
    } else {
      return (
        <form
          onSubmit={this.handleSubmit}
          noValidate
          className={this.props.className}
        >
          <TextInput 
            type="text"
            placeholder="First Name"
            name="firstname"
            handleChange={this.handleChange}
            error={!firstname.valid && firstname.requiredTxt}
          />
          <TextInput 
            type="text"
            placeholder="Last Name"
            name="lastname"
            handleChange={this.handleChange}
            error={!lastname.valid && lastname.requiredTxt}
          />
          <TextInput 
            type="email"
            placeholder="Email"
            name="email"
            handleChange={this.handleChange}
            error={isMatch || isRequired}
          />
          <DatePicker 
            selectedDate={this.state.currentDate}
            onDateClick={this.onDateClick}
            value={datefield.value && format(datefield.value, INPUT_DATE_FORMAT)}
            error={!datefield.valid && datefield.requiredTxt}
            name="datefield"
          /> 
          <Button>Send</Button>
        </form>
      )
    }
  }
}

export const mapStateToProps = (state, ownProps) => {
  const { formReducer: form } = state;
  return {
    message: form && form.data.message,
    status: form && form.data.success
  }
};

export const mapDispatchToProps = (dispatch) => {
  return {
    actions: bindActionCreators(actions, dispatch)
  }
};

EventForm.propTypes = {
  actions: PropTypes.object,
  className: PropTypes.string,
  status: PropTypes.string,
  message: PropTypes.string,
};

export default connect(mapStateToProps, mapDispatchToProps)(EventForm)
