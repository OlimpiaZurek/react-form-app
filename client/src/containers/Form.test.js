import React from 'react';
import { shallow, mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import { EventForm } from './Form';
import ValidationMessage from '../components/ValidationMessage/ValidationMessage';
import TextInput from '../components/TextInput/TextInput';
import DatePicker from '../components/DatePicker/DatePicker';
import { mapStateToProps, mapDispatchToProps } from './Form';
import * as actionTypes from '../actions/actionTypes';
import { data } from '../common/mockData/mockData';

const mockStore = configureStore({});

const props = {
  actions: {submitFormData: () => jest.fn() },
  message: '',
};

describe('<EventForm />', () => {
  let store;
  let wrapper;
  let wrapperInstance;

  beforeEach(() => {
    store = mockStore({});
    wrapper = shallow(
      <EventForm  {...props} store={store}/>
    );
    wrapperInstance = wrapper.instance();
  });

  afterEach(() => {
    wrapper = undefined;
    wrapperInstance = undefined;
  })

  it('should render with given state from Redux store', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a form', () => {
    expect(wrapper.first().type()).toBe('form');
  });

  it('should contains everything else that gets rendered', () => {
    const form = () => wrapper.first();
    expect(form().children()).toEqual(wrapper.children());
  });

  it('should render <ValidationMessage />', () => {
    wrapper.setState({allFieldsValid: true});
    expect(wrapper.find(ValidationMessage).length).toBe(1);
  });

  it('should render <TextInput />', () => {
    expect(wrapper.find(TextInput).length).toBe(3);
  });

  it('should render <DatePicker />', () => {
    expect(wrapper.find(DatePicker).length).toBe(1);
  });

  it('should receives message props as ValidationMessage Text Prop', () => {
    wrapper.setState({allFieldsValid: true});
    const validationMessage =  () => wrapper.find(ValidationMessage);
    expect(validationMessage().props('text')).toEqual(
      wrapper.props('message')
    );
  });

  it('should receives state.valid as <TextInput />  error prop', () => {
    const textInput = () => wrapper.find(TextInput).first();
    expect(textInput().prop('error')).toEqual(false)
  });

  it('should receives handleChange as handleChange <TextInput /> prop', () => {
    const textInput = () => wrapper.find(TextInput).first();
    expect(textInput().prop('handleChange')).toEqual(
      wrapperInstance.handleChange,
    );
  });

  it('should receives state currentDate as <DatePicker />', () => {
    const datepicker = () => wrapper.find(DatePicker).first();
    expect(datepicker().prop('selectedDate')).toEqual(wrapper.state('currentDate'));
  });

  it('should receives onDateClick as onDateClick <DatePicker /> props', () => {
    const datepicker = () => wrapper.find(DatePicker);
    expect(datepicker().prop('onDateClick')).toEqual(
      wrapperInstance.onDateClick
    );
  });

  it('should check valid email when trying to submit empty input ', () => {
    const wrapper = mount(<EventForm {...props}  store={store} />)
    const form = wrapper.find('form');
    form.simulate('submit');
    expect(wrapper.state().email.valid).toBe(false)
  });

  it(' it should return state message from mapStateToProps', () => {
      const state = {
        formReducer: {
          data: {
            message: 'test'
          }
        }
      };
      expect(mapStateToProps(state).message).toEqual('test')
  });

  it('should dispatch actions with mapDispatchToProps', () => {
    const dispatch = jest.fn();
    mapDispatchToProps(dispatch).actions.sendFormDataAction();
    expect(dispatch.mock.calls[0][0]).toEqual({ type: actionTypes.SEND_FORM_DATA});
  });
});
