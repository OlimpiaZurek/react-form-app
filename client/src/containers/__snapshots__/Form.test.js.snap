// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EventForm /> should render with given state from Redux store 1`] = `ShallowWrapper {}`;
